### ISSUE TITLE
Direct Buffer not freed and Netty Resource Leaks 

### ISSUE THREAD

**Issue by mgmeiner** [spring-projects/spring-framework#28382]
First of all: I'm not sure if this is 100% the right place for this Issue as Netty, Reactor-Netty and Spring-Data mongodb are also involved.

After updating our Spring-Webflux app to 2.6.6 we encountered "java.lang.OutOfMemoryError Direct buffer memory" exceptions in our logs. After some investigation I found out that the netty update to 4.1.75/4.1.76 might cause the problem.

When netty leak detection is set to _paranoid_ we see the following leak report in our logs:

```
LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
#1:
        io.netty.buffer.AdvancedLeakAwareByteBuf.writeBytes(AdvancedLeakAwareByteBuf.java:611)
        org.springframework.core.io.buffer.NettyDataBuffer.write(NettyDataBuffer.java:177)
        org.springframework.core.io.buffer.NettyDataBuffer.write(NettyDataBuffer.java:43)
        org.springframework.http.codec.json.AbstractJackson2Encoder.encodeStreamingValue(AbstractJackson2Encoder.java:286)
        org.springframework.http.codec.json.AbstractJackson2Encoder.lambda$encode$1(AbstractJackson2Encoder.java:168)
        reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:106)
...
```
[(complete one attached as file)](https://github.com/spring-projects/spring-framework/files/8562543/leak-log.txt)

The use case behind is: Streaming pretty large data as json (~70kb per json) from mongodb to the client using a spring-webflux application.

Before the update to Spring-Boot 2.6.6. (and the including netty update) the leak was also present in the logs but did not make any problems so far. When starting the app on Java 11 it also seems that the buffer is cleaned fine (probably by GC) - but using Java 17 we always see the  "java.lang.OutOfMemoryError Direct buffer memory" exceptions.

Our App runs on Cloudfoundry with MaxDirectMemorySize set to 80M which before the update was always enough. Increasins the MaxDirectMemorySize just delays the overflow exceptions.

### Example to Reproduce the Issue:

[Here you find a small minimal example but is enough to reproduce the issue: webflux-buffer-overflow-demo](https://github.com/mgmeiner/webflux-buffer-overflow-demo)

Look at the readme of the repository for prerequisites and launch instructions.

For watching the direct buffer pools i used JVisualVM with the Buffer Pools plugins installed.

**Comment by mgmeiner** [comment]
I do not think that the netty update has something to do with it anymore - also i guessed it for a long time but I'm also able to reproduce the buffer 'overflow' when using for example Spring-Boot Version 2.5.7 and MaxDirectMemorySize=128M. 
When MaxDirectMemorySize set to 80M it works - this is so weird. JRE 17 is used.

**Comment by SubhamSinghal** [comment]
Is there any update on this? I am also facing same issue.
spring boot version: 2.6.13
Java version: 17

**Comment by dzmitrykashlach** [comment]
> 2.6.13 Java version: 17
> I have provided 1024M for direct buffer and still getting error. -XX:MaxDirectMemorySize=1024M

Confirm this with java 21, spring-boot 3.3.5, `-XX:MaxDirectMemorySize=80M`

**Comment by Dieken** [comment]
SpringBoot 3.4.4:

```
LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
        io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:410)
        io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
        io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:179)
        io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:116)
        org.springframework.core.io.buffer.NettyDataBufferFactory.allocateBuffer(NettyDataBufferFactory.java:72)
        org.springframework.core.io.buffer.NettyDataBufferFactory.allocateBuffer(NettyDataBufferFactory.java:39)
        org.springframework.http.codec.json.AbstractJackson2Encoder.encodeValue(AbstractJackson2Encoder.java:267)
        org.springframework.http.codec.json.AbstractJackson2Encoder.lambda$encode$0(AbstractJackson2Encoder.java:159)
        reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:106)
        reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:79)
        reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:129)
```

**Comment by Dieken** [comment]
https://github.com/spring-projects/spring-framework/blob/v6.2.6/spring-web/src/main/java/org/springframework/http/codec/EncoderHttpMessageWriter.java#L127

The `body` isn't released under some unknown condition,  notify people who touched this file: @rstoyanchev  @sdeleuze  @jhoeller  @sbrannen @bclozel  @poutsma

**Comment by Dieken** [comment]
> SpringBoot 3.4.4:

Sorry, it's false alarm,  I figure out that's caused by a customized web filter.
